### PR TITLE
Fix segfault when mpi4py is imported before NEST

### LINKED ISSUE #1703: Segmentation fault when calling nest.Simulate() after importing mpi4py
**Describe the bug**
Segmentation fault if I call nest.Simulate after importing mpi4py

**To Reproduce**
```
$ python
>>>from mpi4py import MPI
>>>import nest
>>>nest.Simulate(10)

Jul 29 12:48:38 NodeManager::prepare_nodes [Info]: 
    Preparing 0 nodes for simulation.

Jul 29 12:48:38 MUSICManager::enter_runtime [Info]: 
    Entering MUSIC runtime with tick = 0.1 ms
[etna:07475] *** Process received signal ***
[etna:07475] Signal: Segmentation fault (11)
[etna:07475] Signal code: Address not mapped (1)
[etna:07475] Failing at address: 0x59
[etna:07475] [ 0] /lib/x86_64-linux-gnu/libc.so.6(+0x3ef20)[0x7f5a8c5eff20]
[etna:07475] [ 1] /home/xis/HBP/MUSIC-1.1.16/lib/libmusic.so.1(_ZN5MUSIC5Setup19maybePostponedSetupEv+0x0)[0x7f5a52e4ea70]
[etna:07475] [ 2] /home/xis/HBP/MUSIC-1.1.16/lib/libmusic.so.1(_ZN5MUSIC7RuntimeC1EPNS_5SetupEd+0x1aa)[0x7f5a52e475ca]
[etna:07475] [ 3] /home/xis/HBP/nest-2.18.0/lib/python3.6/site-packages/nest/../../../libnestkernel.so(_ZN4nest12MUSICManager13enter_runtimeEd+0xbe7)[0x7f5a53bed997]
[etna:07475] [ 4] /home/xis/HBP/nest-2.18.0/lib/python3.6/site-packages/nest/../../../libnestkernel.so(_ZN4nest17SimulationManager7prepareEv+0x15b)[0x7f5a53c03edb]
[etna:07475] [ 5] /home/xis/HBP/nest-2.18.0/lib/python3.6/site-packages/nest/../../../libnestkernel.so(_ZN4nest17SimulationManager8simulateERKNS_4TimeE+0x11)[0x7f5a53c0c9c1]
[etna:07475] [ 6] /home/xis/HBP/nest-2.18.0/lib/python3.6/site-packages/nest/../../../libnestkernel.so(_ZN4nest8simulateERKd+0xc2)[0x7f5a53bf19a2]
[etna:07475] [ 7] /home/xis/HBP/nest-2.18.0/lib/python3.6/site-packages/nest/../../../libnestkernel.so(_ZNK4nest10NestModule16SimulateFunction7executeEP14SLIInterpreter+0x43)[0x7f5a53bbda83]
[etna:07475] [ 8] /home/xis/HBP/nest-2.18.0/lib/python3.6/site-packages/nest/../../../libsli.so(_ZN13FunctionDatum7executeEP14SLIInterpreter+0x43)[0x7f5a5360f053]
[etna:07475] [ 9] /home/xis/HBP/nest-2.18.0/lib/python3.6/site-packages/nest/../../../libsli.so(_ZN14SLIInterpreter8execute_Em+0x222)[0x7f5a5360cd32]
[etna:07475] [10] /home/xis/HBP/nest-2.18.0/lib/python3.6/site-packages/nest/../../../libsli.so(_ZN14SLIInterpreter7executeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x15e)[0x7f5a5360d2ae]
[etna:07475] [11] /home/xis/HBP/nest-2.18.0/lib/python3.6/site-packages/nest/pynestkernel.so(+0x37d54)[0x7f5a552ffd54]
[etna:07475] [12] /home/xis/.pyenv/versions/bsb/bin/python[0x50a35c]
[etna:07475] [13] /home/xis/.pyenv/versions/bsb/bin/python(_PyEval_EvalFrameDefault+0x444)[0x50bfb4]
[etna:07475] [14] /home/xis/.pyenv/versions/bsb/bin/python[0x509758]
[etna:07475] [15] /home/xis/.pyenv/versions/bsb/bin/python[0x50a48d]
[etna:07475] [16] /home/xis/.pyenv/versions/bsb/bin/python(_PyEval_EvalFrameDefault+0x444)[0x50bfb4]
[etna:07475] [17] /home/xis/.pyenv/versions/bsb/bin/python[0x507d64]
[etna:07475] [18] /home/xis/.pyenv/versions/bsb/bin/python[0x588d41]
[etna:07475] [19] /home/xis/.pyenv/versions/bsb/bin/python(PyObject_Call+0x3e)[0x59fc4e]
[etna:07475] [20] /home/xis/.pyenv/versions/bsb/bin/python(_PyEval_EvalFrameDefault+0x17e6)[0x50d356]
[etna:07475] [21] /home/xis/.pyenv/versions/bsb/bin/python[0x507d64]
[etna:07475] [22] /home/xis/.pyenv/versions/bsb/bin/python[0x509a90]
[etna:07475] [23] /home/xis/.pyenv/versions/bsb/bin/python[0x50a48d]
[etna:07475] [24] /home/xis/.pyenv/versions/bsb/bin/python(_PyEval_EvalFrameDefault+0x444)[0x50bfb4]
[etna:07475] [25] /home/xis/.pyenv/versions/bsb/bin/python[0x507d64]
[etna:07475] [26] /home/xis/.pyenv/versions/bsb/bin/python(PyEval_EvalCode+0x23)[0x50ae13]
[etna:07475] [27] /home/xis/.pyenv/versions/bsb/bin/python[0x634c82]
[etna:07475] [28] /home/xis/.pyenv/versions/bsb/bin/python[0x4ad695]
[etna:07475] [29] /home/xis/.pyenv/versions/bsb/bin/python(PyRun_InteractiveLoopFlags+0xd4)[0x4afac3]
[etna:07475] *** End of error message ***
Segmentation fault (core dumped)

```

**Desktop/Environment (please complete the following information):**
 - Ubuntu 16.04
 - Python 3.6.9
 - nest-2.18
 - Installation: with MPI, OpenMP and MUSIC as follows:
```
cmake -Dwith-mpi=ON -Dwith-openmp=ON -Dwith-music=ON -DMUSIC_ROOT_DIR=/home/xis/HBP/MUSIC-1.1.16 -DCMAKE_INSTALL_PREFIX:PATH=/home/xis/HBP/nest-2.18.0 /home/xis/HBP/nest-simulator-2.18.0/
```

**Additional context**
Nest, mpi4py and MUSIC are installed in a pyenv environement. Installation of mpi4py works, and it links to a system-level openmpi installation.

Installation:
```
pip install --global-option=build_ext --global-option="-I/usr/lib/x86_64-linux-gnu/openmpi/include/" --global-option="-L/usr/lib/x86_64-linux-gnu/openmpi/lib" mpi4py
```
Test:
```
$ mpiexec -n 5 python -m mpi4py.bench helloworld
Hello, World! I am process 0 of 5 on etna.
Hello, World! I am process 1 of 5 on etna.
Hello, World! I am process 2 of 5 on etna.
Hello, World! I am process 3 of 5 on etna.
Hello, World! I am process 4 of 5 on etna.
```

### PR DESCRIPTION
**Pull request by hakonsbm**
If mpi4py is imported before NEST, `MPI_Init()` is called before NEST is initialized. With MPI already initialized, the communicator is never set in NEST, so any MPI communication, such as when simulating, may end in a segfault.

To avoid a segfault, this PR changes NEST initialization to set the communicator to `MPI_COMM_WORLD` automatically if MPI is already initialized. If NEST is installed with MUSIC support, this is not possible, and an error is issued.

Resolves #1703.

### CODE REVIEW

File: testsuite/regressiontests/issue-1703.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+EXIT_SKIPPED = 200
+EXIT_FAILURE = 127
+EXIT_SEGFAULT = 139
+
+try:
+    import nest
+    from mpi4py import MPI
+except ImportError:
+    # Skip test if mpi4py is not installed, or if NEST is
+    # installed without Python support.
```
Comment (heplesser) [inline testsuite/regressiontests/issue-1703.py:43]: I would move this comment before the try and explain the logic more clearly: here, you test that nest and mpi4py are available (and loading mpi4py after nest is not the problem we are regression-testing here), below comes the actual test. I got at first confused by the try block.

File: testsuite/regressiontests/issue-1703.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+"""
+This script ensures that importing NEST after mpi4py does not lead
+to a segmentation fault when simulating.
+
+This is a regression test for GitHub issue 1703.
+"""
+
+import sys
+import subprocess as sp
```
Comment (heplesser) [inline testsuite/regressiontests/issue-1703.py:30]: I'd suggest to drop the `as sp` for clarity below.

File: nestkernel/mpi_manager.cpp [hunk 3]
```diff
@@ -120,9 +120,19 @@ nest::MPIManager::init_mpi( int* argc, char** argv[] )
     set_communicator( MPI_COMM_WORLD );
 #endif /* #ifdef HAVE_MUSIC */
   }
```
Comment (heplesser) [inline nestkernel/mpi_manager.cpp:122]: On line 116 above is a C-style cast to `MPI_Comm`. Could you replace that with a proper `static_cast`? (Minor tidying up should be allowed in PRs ;)).

**Review by heplesser — COMMENTED** [review]
@hakonsbm Thanks! Just a few small comments.

**Review by heplesser — APPROVED** [review]

**Comment by hakonsbm** [comment]
Looks like the regressiontest hangs on Travis when NEST is compiled with MUSIC, but I can't reproduce it on my machine. Running installcheck without eating the output [shows](https://travis-ci.org/github/hakonsbm/nest-simulator/jobs/720933074#L7756) that it probably hangs on `MPI_Abort()` without actually aborting.

File: testsuite/regressiontests/issue-1703.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+# If not, the test is skipped.
+try:
+    import nest
+    from mpi4py import MPI
+except ImportError:
+    sys.exit(EXIT_SKIPPED)
+
+# Attempt to import MPI from mpi4py before NEST. Running the script in a separate process,
+# in case it ends in a segmentation fault.
+cmd = shlex.split('python -c "from mpi4py import MPI; import nest; nest.Simulate(10)"')
```
Comment (heplesser) [inline testsuite/regressiontests/issue-1703.py:48]: Are we certain that this `python` will invoke the correct Python instance?

**Comment by heplesser** [comment]
I wonder about thre things:

1. Could it be that we invoke a different Python and thus MPI4Py instance than expected and this causes problems?
1. Could it be that Travis does something to `SIGTERM` signals that creates problems?

Maybe @lekshmideepu or @jougs have thoughts on this?

File: testsuite/regressiontests/issue-1703.py (hunk 4 above)
Comment (jougs) [reply testsuite/regressiontests/issue-1703.py:48]: I also think that this should be `python3` instead of just `python`.

**Comment by jougs** [comment]
I don't know about any `SIGTERM` manipulation, but also don't think that there is one.

**Comment by lekshmideepu** [comment]
@heplesser I am also not aware of any problems with SIGTERM under Travis. I could may be take a look into the details.

File: testsuite/regressiontests/issue-1703.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    if exit_code == 1:
+        sys.exit(EXIT_SUCCESS)
+    elif exit_code == -11:
+        sys.exit(EXIT_SEGFAULT)
+    else:
+        sys.exit(EXIT_FAILURE)
+else:
+    if exit_code == 0:
+        sys.exit(EXIT_SUCCESS)
+    elif exit_code == -11:
```
Comment (gtrensch) [inline testsuite/regressiontests/issue-1703.py:62]: Is there a reason for the raw numbers in the code (lines 53, 55 ..)? Using constants and descriptive names would improve readability.

Comment (hakonsbm) [reply testsuite/regressiontests/issue-1703.py]: Added!

**Review by gtrensch — COMMENTED** [review]
The source code changes looks good to me now!
There is still the failing OSX Travis build that hangs on the regression test.

**Comment by hakonsbm** [comment]
Passing the environment to the `subprocess` call fixed the problems Travis had.

**Review by jougs — APPROVED** [review]
The concerns by @gtrensch were resolved. I'm thus formally approving. Many thanks!

**Pull request merged**